File: Backend/BooksAPI/Services/Book/BookService.cs
using BooksAPI.Data;
using BooksAPI.Models;
using Microsoft.EntityFrameworkCore;
using System.Net.Http.Headers;

namespace BooksAPI.Services
{
    public class BookService : IBookService
    {
        private readonly AppDBContext _context;

        public BookService(AppDBContext context)
        {
            _context = context;
        }

        public async Task<bool> Create(BookModel model)
        {
            _context.Books.Add(model);
            {
                await _context.SaveChangesAsync();
                return true;
            }
        }

        public async Task<bool> Delete(Guid id)
        {
            var book = await _context.Books.FindAsync(id);

            if (book == null)
            {
                return false;
            }

            _context.Books.Remove(book);

            await _context.SaveChangesAsync();

            return true;
        }

        public async Task<BookModel> Get(Guid id)
        {
            var book = await _context.Books
                .Include(x => x.Category)
                .Include(y => y.User)
                .FirstOrDefaultAsync(i => i.Id == id);

            return book;
        }

        public async Task<List<BookModel>> GetAll()
        {
            var result = await _context.Books
                .Include(x => x.Category)
                .Include(y => y.User)
                .ToListAsync();

            return result;
        }

        public async Task<List<BookModel>> GetAllByUserAndStatus(string userId, bool isSold)
        {
            var result = await _context.Books
                .Where(c => c.UserId == userId)
                .Where(x => x.isSold == isSold)
                .Include(x => x.Category)
                .Include(y => y.User)
                .ToListAsync();

            return result;
        }

        public async Task<List<BookModel>> GetAllByStatus(bool isSold)
        {
            var result = await _context.Books
                .Where(x => x.isSold == isSold)
                .Include(x => x.Category)
                .Include(y => y.User)
                .ToListAsync();

            return result;
        }

        public async Task<List<BookModel>> GetAllByCategory(string categoryName)
        {
            var result = await _context.Books
                .Where(c => c.Category.Name == categoryName)
                .Include(x => x.Category)
                .Include(y => y.User)
                .ToListAsync();

            return result;
        }

        public Task<bool> Update(Guid id, BookModel model)
        {
            throw new NotImplementedException();
        }

        public async Task<bool> UpdateCertainProperties(Guid id, UpdateBookModel model)
        {
            if (model.Id == null)
            {
                return false;
            }
            var book = await _context.Books.FindAsync(id);

            book.ISBN = model.ISBN;
            book.CategoryId = model.CategoryId;
            book.Title = model.Title;
            book.Author = model.Author;
            book.Publisher = model.Publisher;
            book.PublicationDate = model.PublicationDate;
            book.Page = model.Page;
            book.Language = model.Language;
            book.Condition = model.Condition;
            book.isSold = model.isSold;

            await _context.SaveChangesAsync();

            return true;
        }

        public async Task<bool> UpdatePhoto(Guid id, ImageModel model)
        {
            var book = await _context.Books.FindAsync(id);

            book.Image = model.Image;

            await _context.SaveChangesAsync();

            return true;
        }
    }
}

File: Frontend/src/app/services/book.service.ts
import { HttpClient, HttpEvent, HttpHeaders } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { BehaviorSubject, Observable } from 'rxjs';
import { map, tap } from 'rxjs/operators';
import { Book } from '../models/book';
import { UpdateBook } from '../models/updateBook';
import { Image } from '../models/image';

@Injectable({
  providedIn: 'root',
})
export class BookService {
  readonly baseUrl = 'https://localhost:7295/books';

  readonly httpOptions = {
    headers: new HttpHeaders({
      'Content-Type': 'application/json',
    }),
  };

  getHttpOptionsBearerTextResponse() {
    const httpOptionsJWT = {
      responseType: 'text' as 'json',
      headers: new HttpHeaders({
        Authorization: 'Bearer ' + JSON.parse(localStorage.getItem('token')),
      }),
    };
    return httpOptionsJWT;
  }

  getHttpOptionsBearer() {
    const httpOptionsJWT = {
      headers: new HttpHeaders({
        Authorization: 'Bearer ' + JSON.parse(localStorage.getItem('token')),
      }),
    };
    return httpOptionsJWT;
  }

  readonly httpOptionsText = {
    headers: new HttpHeaders({
      Accept: 'text/html, application/xhtml+xml, */*',
      'Content-Type': 'text/plain, charset=utf-8',
    }),
    observe: 'response',
    responseType: 'text',
  };

  public _searchPropertySubject = new BehaviorSubject<string>('');
  searchProperty$ = this._searchPropertySubject.asObservable();

  private _searchValueSubject = new BehaviorSubject<string>('');
  searchValue$ = this._searchValueSubject.asObservable();

  constructor(private httpClient: HttpClient) {}

  setCurrentProperty(property: string) {
    this._searchPropertySubject.next(property);
  }

  setCurrentValue(value: string) {
    return this._searchValueSubject.next(value);
  }

  getBooks$(): Observable<Book[]> {
    return this.httpClient.get<Book[]>(this.baseUrl, this.httpOptions);
  }

  async getBooksAsync() {
    return await this.httpClient
      .get<Book[]>(this.baseUrl, this.httpOptions)
      .toPromise();
  }

  async getBooksByUserAndStatusAsync$(userId: string, isSold: boolean) {
    return await this.httpClient
      .get<Book[]>(this.baseUrl + '/' + userId + '/' + isSold, this.httpOptions)
      .toPromise();
  }

  async getBooksByStatusAsync(isSold: boolean) {
    return await this.httpClient
      .get<Book[]>(this.baseUrl + '/status/' + isSold, this.httpOptions)
      .toPromise();
  }

  getBooksByStatus$(isSold: boolean): Observable<Book[]> {
    return this.httpClient.get<Book[]>(
      this.baseUrl + '/status/' + isSold,
      this.httpOptions
    );
  }
  async getBooksByCategoryAsync$(categoryName: string) {
    return await this.httpClient
      .get<Book[]>(this.baseUrl + '/category/' + categoryName, this.httpOptions)
      .toPromise();
  }

  getSearchedBooks$(title: string): Observable<Book[]> {
    return this.httpClient
      .get<Book[]>(this.baseUrl, this.httpOptions)
      .pipe(map((books) => books.filter((book) => book.title.includes(title))));
  }

  getBookById$(id: string): Observable<Book> {
    return this.httpClient.get<Book>(this.baseUrl + '/' + id, this.httpOptions);
  }

  async getBookByIdAsync$(id: string) {
    return await this.httpClient
      .get<Book>(this.baseUrl + '/' + id, this.httpOptions)
      .toPromise();
  }

  addBook$(
    isbn: string,
    // uid: string,
    userId: string,
    title: string,
    author: string,
    publisher: string,
    publicationDate: Date,
    page: number,
    price: number,
    language: string,
    condition: string,
    categoryId: string,
    image: string,
    isSold: boolean
  ): Observable<Book> {
    let book = {
      isbn: isbn,
      userId: userId,
      title: title,
      author: author,
      publisher: publisher,
      publicationDate: publicationDate,
      page: page,
      price: price,
      language: language,
      condition: condition,
      categoryId: categoryId,
      image: image,
      isSold: isSold,
    };
    return this.httpClient.post<Book>(
      this.baseUrl,
      book,
      this.getHttpOptionsBearer()
    );
  }

  updateBook$(id: string, model): Observable<UpdateBook> {
    let book = {
      categoryId: model.categoryId,
      isbn: model.isbn,
      title: model.title,
      author: model.author,
      publisher: model.publisher,
      publicationDate: model.publicationDate,
      page: model.page,
      language: model.language,
      condition: model.condition,
      isSold: model.isSold,
    };
    return this.httpClient.put<UpdateBook>(
      this.baseUrl + '/' + id,
      book,
      this.getHttpOptionsBearer()
    );
  }

  saveImageBook$(image: File): Observable<HttpEvent<Response>> {
    const formData = new FormData();

    formData.append('image', image, image.name);

    return this.httpClient.post<Response>(
      this.baseUrl + '/saveFile',
      formData,
      {
        reportProgress: true,
        observe: 'events',
      }
    );
  }

  updatePhotoBook$(id: string, model: Image): Observable<Image> {
    let img: Image = {
      image: model.image,
    };
    return this.httpClient.put<Image>(
      this.baseUrl + '/book/updatePhoto/' + id,
      img,
      this.getHttpOptionsBearer()
    );
  }

  deleteBook(id: string) {
    return this.httpClient.delete<Book>(
      this.baseUrl + '/' + id,
      this.getHttpOptionsBearerTextResponse()
    );
  }
}
